Build fails with `g++` 11.2.0 (`-Werror=maybe-uninitialized`)

### Describe the bug

```
$ make RELEASE=1
…
g++  -DRELEASE -DGIT_VERSION -DBACKTRACE -ffast-math -Os  -Werror -Wall -Wextra -Wformat-signedness -Wlogical-op -Wmissing-declarations -Wmissing-noreturn -Wnon-virtual-dtor -Wold-style-cast -Woverloaded-virtual -Wpedantic -Wsuggest-override -Wunused-macros -Wzero-as-null-pointer-constant -Wno-unknown-warning-option -Wno-range-loop-analysis  -Wredundant-decls    -fsigned-char  -std=c++14 -MMD -MP -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -Ipch -Winvalid-pch -fpch-preprocess -include main-pch.hpp -c src/activity_item_handling.cpp -o obj/activity_item_handling.o
src/activity_item_handling.cpp: In function ‘activity_reason_info find_base_construction(const std::vector<construction>&, player&, const inventory&, const tripoint&, const cata::optional<int_id<construction> >&, construction_id, std::set<int_id<construction> >&)’:
src/activity_item_handling.cpp:1168:73: error: ‘*(int*)((char*)&reason + offsetof(cata::optional<do_activity_reason>,cata::optional<do_activity_reason>::<unnamed>))’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
 1168 |         return activity_reason_info::build( *reason, false, pre_req_idx );
      |                                                                         ^
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-range-loop-analysis’ may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
make: *** [Makefile:887: obj/activity_item_handling.o] Error 1
``` 

If warnings are disabled by uncommenting `WARNINGS = -w` in the Makefile, I seem to get a working build.

### Versions and configuration

- Ubuntu 22.04, `g++` 11.2.0
- f5cccb87dfe69c436b5e5f82fc9efd4a91e1e33b

--- a/src/effect.cpp
+++ b/src/effect.cpp
@@ -1384,8 +1384,8 @@ void load_effect_type( const JsonObject &jo )
     }
 
     // TODO: Implement handling of reduced morale, remove this
-    static const std::vector<const char *> mod_types = {{
-            "base_mods", "scaling_mods"
+    static const std::vector<std::string> mod_types = {{
+            {"base_mods"}, {"scaling_mods"}
         }
     };
     if( has_morale_effect ) {

--- a/src/optional.h
+++ b/src/optional.h
@@ -32,6 +32,9 @@ class optional
     private:
         using StoredType = typename std::remove_const<T>::type;
         union {
+            // `volatile` suppresses -Wmaybe-uninitialized false positive
+            // See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635#c53
+            volatile char dont_use;
             char dummy;
             StoredType data;
         };

--- a/src/wisheffect.cpp
+++ b/src/wisheffect.cpp
@@ -37,7 +37,7 @@ struct entry_data {
     cata::optional<const effect *> eff;
 };
 
-template<class T, typename F = std::pair<std::string, void ( T::* )(
+template<class T, typename F = std::pair<const std::string, void ( T::* )(
              const input_context &,
              const input_event &,
              int,

